federation/ipatuura/src/main/java/org/keycloak/ipatuura_user_spi/IpatuuraUserStorageProvider.java

```diff
@@ -0,0 +1,329 @@
+import org.keycloak.models.UserModel;
+import org.keycloak.models.credential.PasswordCredentialModel;
+import org.keycloak.storage.StorageId;
+import org.keycloak.storage.UserStorageProvider;
+import org.keycloak.storage.UserStoragePrivateUtil;
+import org.keycloak.storage.user.ImportedUserValidation;
+import org.keycloak.storage.user.UserLookupProvider;
+import org.keycloak.storage.user.UserRegistrationProvider;
+import org.keycloak.broker.provider.util.SimpleHttp;
+
+import org.keycloak.ipatuura_user_spi.authenticator.IpatuuraAuthenticator;
+import org.keycloak.ipatuura_user_spi.schemas.SCIMError;
+import org.keycloak.ipatuura_user_spi.schemas.SCIMUser;
+
+import org.keycloak.storage.user.UserQueryProvider;
+
+import java.io.IOException;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Stream;
+
+import org.apache.http.HttpStatus;
+
+/**
+ * @author <a href="mailto:jstephen@redhat.com">Justin Stephenson</a>
+ * @version $Revision: 1 $
+ */
+public class IpatuuraUserStorageProvider implements UserStorageProvider, UserLookupProvider, CredentialInputValidator,
+        CredentialAuthentication, UserRegistrationProvider, UserQueryProvider, ImportedUserValidation {
+    protected KeycloakSession session;
+    protected ComponentModel model;
+    protected Ipatuura ipatuura;
+    private static final Logger logger = Logger.getLogger(IpatuuraUserStorageProvider.class);
+    protected final Set<String> supportedCredentialTypes = new HashSet<>();
+    protected IpatuuraUserStorageProviderFactory factory;
+
+    public IpatuuraUserStorageProvider(KeycloakSession session, ComponentModel model, Ipatuura ipatuura,
+                                       IpatuuraUserStorageProviderFactory factory) {
+        this.session = session;
+        this.model = model;
+        this.ipatuura = ipatuura;
+        this.factory = factory;
+
+        supportedCredentialTypes.add(PasswordCredentialModel.TYPE);
+    }
+
+    @Override
+    public UserModel getUserByEmail(RealmModel realm, String email) {
+        return null;
+    }
+
+    @Override
+    public UserModel getUserById(RealmModel realm, String id) {
+        StorageId storageId = new StorageId(id);
+        String username = storageId.getExternalId();
+        return getUserByUsername(realm, username);
+    }
+
+    @Override
+    public UserModel getUserByUsername(RealmModel realm, String username) {
+        /*
+         * Remove @realm, this is needed as GSSAPI auth users reach here as user@realm
+         */
+        int idx = username.indexOf("@");
+        if (idx != -1) {
+            username = username.substring(0, idx);
+        }
+
+        UserModel user = UserStoragePrivateUtil.userLocalStorage(session).getUserByUsername(realm, username);
+        if (user != null) {
+            logger.debug("User already exists in keycloak");
+            return user;
```

Comment:
We should probably consider in the next iteration to return a delegate that tracks state changes to sync them back to the SCIM service provider in a transaction.

Btw, do we want to return `IpatuuraUserModelDelegate` instead when the local account exists?

Reply:
I'm not sure if the code ever reaches this - the call `getUserByUsername` goes through the `UserStorageManager`, which checks the local storage and calls `validate` on the provider. This method is returning the delegate already.

Reply:
So in a way the provider is repeating the local storage check needlessly